clang/lib/Parse/ParseContracts.cpp

```diff
@@ -0,0 +1,437 @@
+  bool IsInvalidTmp = false;
+  return ParseFunctionContractSpecifierImpl({}, CSO_FunctionContext, IsInvalidTmp);
+}
+
+/// ParseFunctionContractSpecifierSeq - Parse a series of pre/post contracts on
+/// a function declaration.
+///
+///   function-contract-specifier-seq :
+///       function-contract-specifier function-contract-specifier-seq
+//
+///   function-contract-specifier:
+///       precondition-specifier
+///       postcondition-specifier
+//
+///   precondition-specifier:
+///       pre attribute-specifier-seq[opt] ( conditional-expression )
+///
+///   postcondition-specifier:
+///       post attribute-specifier-seq[opt] ( result-name-introducer[opt]
+///       conditional-expression )
+///
+///   result-name-introducer:
+///       attributed-identifier :
+void Parser::ParseContractSpecifierSequence(Declarator &DeclarationInfo,
+                                            bool EnterScope,
+                                            QualType TrailingReturnType) {
+  if (!isFunctionContractKeyword(Tok))
+    return;
+
+  std::optional<QualType> CachedType;
+  auto ReturnTypeResolver = [&]() {
+    if (!CachedType) {
+      QualType ReturnType = TrailingReturnType;
+      if (ReturnType.isNull()) {
+        TypeSourceInfo *TInfo = Actions.GetTypeForDeclarator(DeclarationInfo);
+        assert(TInfo && TInfo->getType()->isFunctionType());
+        ReturnType = TInfo->getType()->getAs<FunctionType>()->getReturnType();
+      }
+      CachedType = ReturnType;
+    }
+    return CachedType.value();
+  };
+  std::optional<ParseScope> ParserScope;
+
+  std::optional<Sema::CXXThisScopeRAII> ThisScope;
+  std::optional<Sema::FunctionScopeRAII> PopFnContext;
+
+  if (EnterScope) {
+    ParserScope.emplace(this, Scope::DeclScope | Scope::FunctionPrototypeScope |
+                                  Scope::FunctionDeclarationScope);
+
+   // PopFnContext.emplace(Actions);
+   // Actions.PushFunctionScope();
+
+    auto FTI = DeclarationInfo.getFunctionTypeInfo();
+
+    for (unsigned i = 0; i != FTI.NumParams; ++i) {
+      ParmVarDecl *Param = cast<ParmVarDecl>(FTI.Params[i].Param);
+      Actions.ActOnReenterCXXMethodParameter(getCurScope(), Param);
+    }
+  }
+
+  InitCXXThisScopeForDeclaratorIfRelevant(DeclarationInfo,
+                                          DeclarationInfo.getDeclSpec(),
+                                          ThisScope);
+  bool IsInvalid = false;
+  SourceLocation StartLoc = Tok.getLocation();
+
+  SmallVector<ContractStmt *, 4> Contracts;
+  while (isFunctionContractKeyword(Tok)) {
+    bool IsInvalidTmp = false;
+    StmtResult Contract =
+        ParseFunctionContractSpecifierImpl(ReturnTypeResolver, EnterScope ? CSO_ParentContext : CSO_FunctionContext, IsInvalidTmp);
+    IsInvalid |= IsInvalidTmp;
+    if (Contract.isUsable())
+      Contracts.push_back(Contract.getAs<ContractStmt>());
+  }
+  ContractSpecifierDecl *Seq = Actions.ActOnFinishContractSpecifierSequence(
+      Contracts, StartLoc, IsInvalid);
+
```

Comment:
You should check we are parsing a function (and not a type or a function pointer)
see `err_requires_clause_on_declarator_not_declaring_a_function`